BUG: unused function warnings in current build

When I build the current `main` branch using clang on an ARM mac, I get the following warnings:

```
[119/314] Compiling C object numpy/_core/l...generated_loops_trigonometric.dispatch.c.o
../numpy/_core/src/umath/loops_trigonometric.dispatch.c.src:168:1: warning: unused function 'simd_cos_f64' [-Wunused-function]
simd_cos_f64(const double *src, npy_intp ssrc, double *dst, npy_intp sdst, npy_intp len)
^
../numpy/_core/src/umath/loops_trigonometric.dispatch.c.src:168:1: warning: unused function 'simd_sin_f64' [-Wunused-function]
simd_sin_f64(const double *src, npy_intp ssrc, double *dst, npy_intp sdst, npy_intp len)
^
2 warnings generated.
[136/314] Compiling C object numpy/_core/l...on-generated_loops_arithmetic.dispatch.c.o
../numpy/_core/src/umath/loops_arithmetic.dispatch.c.src:46:1: warning: unused function 'simd_divide_by_scalar_contig_s64' [-Wunused-function]
simd_divide_by_scalar_contig_s64(char **args, npy_intp len)
^
../numpy/_core/src/umath/loops_arithmetic.dispatch.c.src:112:1: warning: unused function 'simd_divide_by_scalar_contig_u64' [-Wunused-function]
simd_divide_by_scalar_contig_u64(char **args, npy_intp len)
^
2 warnings generated.
[253/314] Compiling C++ object numpy/linal...cpython-311-darwin.so.p/umath_linalg.cpp.o
../numpy/linalg/umath_linalg.cpp:2279:1: warning: unused function 'call_geev' [-Wunused-function]
call_geev(GEEV_PARAMS_t<fortran_complex>* params)
^
1 warning generated.
[276/314] Compiling C object numpy/_core/l.../meson-generated_loops_minmax.dispatch.c.o
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_max_f32' [-Wunused-function]
simd_binary_max_f32(const npyv_lanetype_f32 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_min_f32' [-Wunused-function]
simd_binary_min_f32(const npyv_lanetype_f32 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_maxp_f32' [-Wunused-function]
simd_binary_maxp_f32(const npyv_lanetype_f32 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_minp_f32' [-Wunused-function]
simd_binary_minp_f32(const npyv_lanetype_f32 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_max_f64' [-Wunused-function]
simd_binary_max_f64(const npyv_lanetype_f64 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_min_f64' [-Wunused-function]
simd_binary_min_f64(const npyv_lanetype_f64 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_maxp_f64' [-Wunused-function]
simd_binary_maxp_f64(const npyv_lanetype_f64 *ip1, npy_intp sip1,
^
../numpy/_core/src/umath/loops_minmax.dispatch.c.src:230:1: warning: unused function 'simd_binary_minp_f64' [-Wunused-function]
simd_binary_minp_f64(const npyv_lanetype_f64 *ip1, npy_intp sip1,
^
8 warnings generated.
```

gcc doesn't seem to notice these

--- a/numpy/_core/src/umath/loops_arithmetic.dispatch.c.src
+++ b/numpy/_core/src/umath/loops_arithmetic.dispatch.c.src
@@ -36,12 +36,20 @@
  *     q = TRUNC((n - (-dsign ) + (-nsign))/d) - (-qsign);
  ********************************************************************************/
 
+#if (defined(NPY_HAVE_VSX) && !defined(NPY_HAVE_VSX4)) || defined(NPY_HAVE_NEON)
+    // Due to integer 128-bit multiplication emulation, SIMD 64-bit division
+    // may not perform well on both neon and up to VSX3 compared to scalar
+    // division.
+    #define SIMD_DISABLE_DIV64_OPT
+#endif
+
 #if NPY_SIMD
 /**begin repeat
  * Signed types
  * #sfx    = s8, s16, s32, s64#
  * #len    = 8,  16,  32,  64#
  */
+#if @len@ < 64 || (@len@ == 64 && !defined(SIMD_DISABLE_DIV64_OPT))
 static inline void
 simd_divide_by_scalar_contig_@sfx@(char **args, npy_intp len)
 {
@@ -101,6 +109,7 @@ simd_divide_by_scalar_contig_@sfx@(char **args, npy_intp len)
     }
     npyv_cleanup();
 }
+#endif
 /**end repeat**/
 
 /**begin repeat
@@ -108,6 +117,7 @@ simd_divide_by_scalar_contig_@sfx@(char **args, npy_intp len)
  * #sfx    = u8, u16, u32, u64#
  * #len    = 8,  16,  32,  64#
  */
+#if @len@ < 64 || (@len@ == 64 && !defined(SIMD_DISABLE_DIV64_OPT))
 static inline void
 simd_divide_by_scalar_contig_@sfx@(char **args, npy_intp len)
 {
@@ -129,6 +139,7 @@ simd_divide_by_scalar_contig_@sfx@(char **args, npy_intp len)
     }
     npyv_cleanup();
 }
+#endif
 /**end repeat**/
 
 #if defined(NPY_HAVE_VSX4)
@@ -335,8 +346,7 @@ vsx4_simd_divide_contig_@sfx@(char **args, npy_intp len)
     #define TO_SIMD_SFX(X) X##_s@len@
 /**end repeat1**/
 #endif
-
-#if NPY_BITSOF_@TYPE@ == 64 && !defined(NPY_HAVE_VSX4) && (defined(NPY_HAVE_VSX) || defined(NPY_HAVE_NEON))
+#if NPY_BITSOF_@TYPE@ == 64 && defined(SIMD_DISABLE_DIV64_OPT)
     #undef TO_SIMD_SFX
 #endif
 

--- a/numpy/_core/src/umath/loops_exponent_log.dispatch.c.src
+++ b/numpy/_core/src/umath/loops_exponent_log.dispatch.c.src
@@ -124,18 +124,6 @@ fma_blend(__m256 x, __m256 y, __m256 ymask)
 }
 
 NPY_FINLINE __m256
-fma_invert_mask_ps(__m256 ymask)
-{
-    return _mm256_andnot_ps(ymask, _mm256_set1_ps(-1.0));
-}
-
-NPY_FINLINE __m256i
-fma_invert_mask_pd(__m256i ymask)
-{
-    return _mm256_andnot_si256(ymask, _mm256_set1_epi32(0xFFFFFFFF));
-}
-
-NPY_FINLINE __m256
 fma_get_exponent(__m256 x)
 {
     /*
@@ -309,18 +297,6 @@ NPY_FINLINE __m512
 avx512_blend(__m512 x, __m512 y, __mmask16 ymask)
 {
     return _mm512_mask_mov_ps(x, ymask, y);
-}
-
-NPY_FINLINE __mmask16
-avx512_invert_mask_ps(__mmask16 ymask)
-{
-    return _mm512_knot(ymask);
-}
-
-NPY_FINLINE __mmask8
-avx512_invert_mask_pd(__mmask8 ymask)
-{
-    return _mm512_knot(ymask);
 }
 
 NPY_FINLINE __m512

--- a/numpy/_core/src/umath/loops_minmax.dispatch.c.src
+++ b/numpy/_core/src/umath/loops_minmax.dispatch.c.src
@@ -225,7 +225,8 @@ simd_binary_ccc_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip1, const npyv_lanety
     }
 }
 // non-contiguous for float 32/64-bit memory access
-#if @is_fp@
+#if @is_fp@ && !defined(NPY_HAVE_NEON)
+// unroll scalars faster than non-contiguous vector load/store on Arm
 static inline void
 simd_binary_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip1, npy_intp sip1,
                            const npyv_lanetype_@sfx@ *ip2, npy_intp sip2,

--- a/numpy/_core/src/umath/loops_trigonometric.dispatch.c.src
+++ b/numpy/_core/src/umath/loops_trigonometric.dispatch.c.src
@@ -19,8 +19,9 @@
 /**begin repeat
  *  #check = F64, F32#
  *  #sfx  = f64, f32#
+ *  #enable = 0, 1#
  */
-#if NPY_SIMD_@check@
+#if NPY_SIMD_@check@ && @enable@
 /*
  * Vectorized Cody-Waite range reduction technique
  * Performs the reduction step x* = x - y*C in three steps:
@@ -39,8 +40,10 @@ simd_range_reduction_@sfx@(npyv_@sfx@ x, npyv_@sfx@ y, npyv_@sfx@ c1, npyv_@sfx@
 }
 #endif
 /**end repeat**/
-
-#if NPY_SIMD_F64
+/* Disable SIMD code and revert to libm: see
+ * https://mail.python.org/archives/list/numpy-discussion@python.org/thread/C6EYZZSR4EWGVKHAZXLE7IBILRMNVK7L/
+ * for detailed discussion on this*/
+#if 0 // NPY_SIMD_F64
 /**begin repeat
  *  #op = cos, sin#
  */

--- a/numpy/_core/src/umath/loops_unary.dispatch.c.src
+++ b/numpy/_core/src/umath/loops_unary.dispatch.c.src
@@ -195,6 +195,8 @@ simd_unary_nc_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip, npy_intp istride,
 #undef UNROLL
 #define UNROLL 2
 #endif
+// X86 does better with unrolled scalar for heavy non-contiguous
+#ifndef NPY_HAVE_SSE2
 static NPY_INLINE void
 simd_unary_nn_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip, npy_intp istride,
                              npyv_lanetype_@sfx@ *op, npy_intp ostride,
@@ -226,6 +228,7 @@ simd_unary_nn_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip, npy_intp istride,
         *op = scalar_@intrin@(*ip);
     }
 }
+#endif // NPY_HAVE_SSE2
 #endif // @supports_ncontig@
 #undef UNROLL
 #endif // @simd_chk@
@@ -314,8 +317,8 @@ NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(@TYPE@_@kind@)
                 );
                 goto clear;
             }
-        // SSE2 does better with unrolled scalar for heavy non-contiguous
-        #if !defined(NPY_HAVE_SSE2)
+        // X86 does better with unrolled scalar for heavy non-contiguous
+        #ifndef NPY_HAVE_SSE2
             else if (istride != 1 && ostride != 1) {
                 // non-contiguous input and output
                 TO_SIMD_SFX(simd_unary_nn_@intrin@)(

--- a/numpy/linalg/umath_linalg.cpp
+++ b/numpy/linalg/umath_linalg.cpp
@@ -2298,7 +2298,7 @@ process_geev_results(GEEV_PARAMS_t<typ> *params, scalar_trait)
     }
 }
 
-
+#if 0
 static inline fortran_int
 call_geev(GEEV_PARAMS_t<fortran_complex>* params)
 {
@@ -2314,6 +2314,8 @@ call_geev(GEEV_PARAMS_t<fortran_complex>* params)
                           &rv);
     return rv;
 }
+#endif
+
 static inline fortran_int
 call_geev(GEEV_PARAMS_t<fortran_doublecomplex>* params)
 {

